### ISSUE TITLE
[New Resource]: Enable Default Host Management Configuration DHMC  

### ISSUE THREAD

**Issue by tazmaniq2** [hashicorp/terraform-provider-aws#30474]
### Description

AWS announced the ability to enable AWS Systems Manager by default across all EC2 instances in an account with a single action, using Default Host Management Configuration (DHMC) but currently is unsupported by Terraform provider.
It would be helpful to add it as a new resource in order to enable it and also as a data source to get the current status.

https://docs.aws.amazon.com/systems-manager/latest/userguide/managed-instances-default-host-management.html

### Requested Resource(s) and/or Data Source(s)

resource "aws_ssm_dhmc" "example" {}

data "aws_ssm_dhmc" "example" {}

### Potential Terraform Configuration

```terraform
resource "aws_ssm_dhmc" "example" {
  enable   = true
  iam_role = aws_iam_role.AWSSystemsManagerDefaultEC2InstanceManagementRole
}

resource "aws_iam_role" "AWSSystemsManagerDefaultEC2InstanceManagementRole" {
  name               = "AWSSystemsManagerDefaultEC2InstanceManagementRole"
  assume_role_policy = templatefile("${path.module}/policies/trust-policy.json", {})
}

resource "aws_iam_role_policy_attachment" "AmazonSSMManagedEC2InstanceDefaultPolicy" {
  policy_arn = "arn:aws:iam::aws:policy/AmazonSSMManagedEC2InstanceDefaultPolicy"
  role       = aws_iam_role.AWSSystemsManagerDefaultEC2InstanceManagementRole
}

data "aws_ssm_dhmc" "example" {}

output "dhmc_status" {
  value = data.aws_ssm_dhmc.status
}

#trust-policy.json
#{
#    "Version":"2012-10-17",
#    "Statement":[
#        {
#            "Sid":"",
#            "Effect":"Allow",
#            "Principal":{
#                "Service":"ssm.amazonaws.com"
#            },
#            "Action":"sts:AssumeRole"
#        }
#    ]
#}
```

### References

AWS announcement:
https://aws.amazon.com/about-aws/whats-new/2023/02/enable-aws-systems-manager-default-all-ec2-instances-account/
DHMC User Guide:
https://docs.aws.amazon.com/systems-manager/latest/userguide/managed-instances-default-host-management.html

### Would you like to implement a fix?

None

**Comment by dkravetz** [comment]
I don't think this is necessary, you can achieve the same effect by creating and referencing the following resources:

```
module "ssm_default_host_management_role" {
  source  = "terraform-aws-modules/iam/aws//modules/iam-assumable-role"
  version = "5.20.0"

  create_role = true

  trusted_role_services = [
    "ssm.amazonaws.com"
  ]

  role_name         = "AWSSystemsManagerDefaultEC2InstanceManagementRole"
  role_requires_mfa = false

  custom_role_policy_arns = [
    "arn:aws:iam::aws:policy/AmazonSSMManagedEC2InstanceDefaultPolicy",
  ]
}

resource "aws_ssm_service_setting" "default_host_management" {
  setting_id    = "arn:aws:ssm:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:servicesetting/ssm/managed-instance/default-ec2-instance-management-role"
  setting_value = "service-role/AWSSystemsManagerDefaultEC2InstanceManagementRole"
}
```

**Comment by dustyketchum** [comment]
I needed to use

```
resource "aws_ssm_service_setting" "default_host_management" {
  setting_id    = "arn:aws:ssm:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:servicesetting/ssm/managed-instance/default-ec2-instance-management-role"
  setting_value = "AWSSystemsManagerDefaultEC2InstanceManagementRole"
}
```

without `service-role/`, it did not work for me until I changed this.

**Comment by rahulsen** [comment]
> I don't think this is necessary, you can achieve the same effect by creating and referencing the following resources:
> 
> ```
> module "ssm_default_host_management_role" {
>   source  = "terraform-aws-modules/iam/aws//modules/iam-assumable-role"
>   version = "5.20.0"
> 
>   create_role = true
> 
>   trusted_role_services = [
>     "ssm.amazonaws.com"
>   ]
> 
>   role_name         = "AWSSystemsManagerDefaultEC2InstanceManagementRole"
>   role_requires_mfa = false
> 
>   custom_role_policy_arns = [
>     "arn:aws:iam::aws:policy/AmazonSSMManagedEC2InstanceDefaultPolicy",
>   ]
> }
> 
> resource "aws_ssm_service_setting" "default_host_management" {
>   setting_id    = "arn:aws:ssm:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:servicesetting/ssm/managed-instance/default-ec2-instance-management-role"
>   setting_value = "service-role/AWSSystemsManagerDefaultEC2InstanceManagementRole"
> }
> ```

How does this "achieve the same effect" ? It is not enabling DHMC as far as i can see. All it is doing is setting the value of  "servicesetting/ssm/managed-instance/default-ec2-instance-management-role" to the role's name that SSM service will be trusted to assume for the instance to be able to register to SSM. If DHMC is not enabled in the first place, this is useless. Correct me if i am wrong.

**Comment by lorengordon** [comment]
> All it is doing is setting the value of "servicesetting/ssm/managed-instance/default-ec2-instance-management-role" to the role's name that SSM service will be trusted to assume for the instance to be able to register to SSM. If DHMC is not enabled in the first place, this is useless. Correct me if i am wrong.

Yep, that is all that toggle does when "enabling" DHMC. It just sets the ssm service setting to the role to use. Apply that config, and the toggle is enabled.

**Comment by rahulsen** [comment]
Ok thanks. That was the gap in my understanding. I tested it and indeed just setting/resetting that value is the DHMC toggle. I wish AWS documentation was explicit about this.

**Comment by rahulsen** [comment]
> > All it is doing is setting the value of "servicesetting/ssm/managed-instance/default-ec2-instance-management-role" to the role's name that SSM service will be trusted to assume for the instance to be able to register to SSM. If DHMC is not enabled in the first place, this is useless. Correct me if i am wrong.
> 
> Yep, that is all that toggle does when "enabling" DHMC. It just sets the ssm service setting to the role to use. Apply that config, and the toggle is enabled.

Ok thanks. That was the gap in my understanding. I tested it and indeed just setting/resetting that value is the DHMC toggle. I wish AWS documentation was explicit about this.

**Comment by lorengordon** [comment]
> I wish AWS documentation was explicit about this.

Absolutely! I only figured it out by digging through the CloudFormation stack and SSM Documents that are created as part of the "quick setup" when [configuring Default Host Management for an Organization](https://docs.aws.amazon.com/systems-manager/latest/userguide/quick-setup-default-host-management-configuration.html)... Definitely an obscure setup! At the end, I was like, "That's it? Ok, well I can do that easily enough without the quick setup or the CloudFormation it relies on..."

**Comment by Ty182** [comment]
I can't figure out how to run this against multiple Regions since DHMC is Region specific. Any ideas?

```
resource "aws_ssm_service_setting" "default_host_management" {
  for_each = { for idx, region in var.regions : region => idx }

  setting_id    = "arn:aws:ssm:${each.key}:${var.accountid}:servicesetting/ssm/managed-instance/default-ec2-instance-management-role"
  setting_value = "AWSSystemsManagerDefaultEC2InstanceManagementRole"
}
```
I turned this into a module so that it iterates through a list of regions e.g., ["us-east-1", "us-east-2"] and tries to enable DHMC. It works the first time e.g., `us-east-1` but subsequent regions fail. I also simplified this into two resources rather than a loop and module but still the same error. 

```
│ Error: updating AWS SSM (Systems Manager) Service Setting (arn:aws:ssm:us-east-2:123456789012:servicesetting/ssm/managed-instance/default-ec2-instance-management-role): ValidationException: Invalid settingId : arn:aws:ssm:us-east-2:123456789012:servicesetting/ssm/managed-instance/default-ec2-instance-management-role.
│       status code: 400, request id: a298ef20-57ba-47f1-a55c-b247e883039e
│ 
│   with module.ssm_configurations.module.ssm_enablement_dev.aws_ssm_service_setting.default_host_management["us-east-2"],
│   on ssm_configuration/ssm_enablement/main.tf line 4, in resource "aws_ssm_service_setting" "default_host_management":
│    4: resource "aws_ssm_service_setting" "default_host_management" {
```

**Comment by lorengordon** [comment]
fyi, although the API doc _claims_ the `setting_id` should be an ARN, it appears it accepts the simplified "name" also, e.g. `/ssm/managed-instance/default-ec2-instance-management-role`

**Comment by lorengordon** [comment]
Went ahead and published a module to make it easier to set up default host management... https://registry.terraform.io/modules/plus3it/tardigrade-ssm-default-host-management/aws/latest

**Comment by rkohanyi** [comment]
> I don't think this is necessary, you can achieve the same effect by creating and referencing the following resources:
> 
> ```
> module "ssm_default_host_management_role" {
>   source  = "terraform-aws-modules/iam/aws//modules/iam-assumable-role"
>   version = "5.20.0"
> 
>   create_role = true
> 
>   trusted_role_services = [
>     "ssm.amazonaws.com"
>   ]
> 
>   role_name         = "AWSSystemsManagerDefaultEC2InstanceManagementRole"
>   role_requires_mfa = false
> 
>   custom_role_policy_arns = [
>     "arn:aws:iam::aws:policy/AmazonSSMManagedEC2InstanceDefaultPolicy",
>   ]
> }
> 
> resource "aws_ssm_service_setting" "default_host_management" {
>   setting_id    = "arn:aws:ssm:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:servicesetting/ssm/managed-instance/default-ec2-instance-management-role"
>   setting_value = "service-role/AWSSystemsManagerDefaultEC2InstanceManagementRole"
> }
> ```

Bumped into this, for me the following worked:

```
module "ssm_default_host_management_role" {
  source  = "terraform-aws-modules/iam/aws//modules/iam-assumable-role"
  version = "5.48.0"

  create_role = true

  trusted_role_services = [
    "ssm.amazonaws.com"
  ]

  role_name         = "AWSSystemsManagerDefaultEC2InstanceManagementRole"
  role_requires_mfa = false

  custom_role_policy_arns = ["arn:aws:iam::aws:policy/AmazonSSMManagedEC2InstanceDefaultPolicy"]
}

resource "aws_ssm_service_setting" "default_host_management" {
  setting_id    = "arn:aws:ssm:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:servicesetting/ssm/managed-instance/default-ec2-instance-management-role"
  setting_value = "role/service-role/AWSSystemsManagerDefaultEC2InstanceManagementRole"
}
```

Differences are
- module version is bumped up to 5.48.0, not sure if it's relevant or not
- used `role/service-role/AWSSystemsManagerDefaultEC2InstanceManagementRole`, this is the exact role name/path that AWS creates when use enable DHMC feature manually - **missing the `role/` prefix worked to some extent, but Fleet Manager didn't pickup instances, connecting to instance through SSM didn't work**

Hope this helps out someone, I spent on this like 2 and a half hours even after finding this solution 😲
